### PR TITLE
Add a dialog to customize run instances

### LINKED ISSUE #522: Allow to specify command-line arguments when running project instances from the editor
*Related to https://github.com/godotengine/godot-proposals/issues/3357.*

**Describe the project you are working on:**
Multiple games and small apps.

**Describe the problem or limitation you are having in your project:**
When running any project, often I will use a "dev.gd" file or other file to change the launch state of the application.  However, doing this is problematic because: 

1 - Finding the "dev.gd" file is frustrating when many GD files are open
2 - Don't want the changes to always be permanent across sessions
3 - Don't want the changes to always be included with a release build (often forget to undo the changes back to blank state, end up building again to wipe debug settings)

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
A simple line text edit, white font on black background.  Optional and must be enabled in Godot Editor Settings.  In this, the user of the editor can set a string for command line arguments to be passed to the application.  For example:

"level=10 cash=50 health=800 scene=end-fight"

The application would be responsible for parsing these arguments just like regular command line arguments.  When the Godot Editor is restarted, this console is blank.  This console does not have permanence, the arguments could not be included in a release build.  Something like these removes a bulk of the need for a constantly changing development GD file, a file which can cause havoc if not reset to blank on release.  Having the arguments always visible and handy in the editor would be a relief.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
The argument string would be passed to the application at execution.  The application would be responsible for using the arguments (would be ignored by default).

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
It would be used very frequently, practically every session.

**Is there a reason why this should be core and not an add-on in the asset library?:**
A lot of developers would love this feature!!  :)

### PR DESCRIPTION
**Pull request by KoBeWi**
Closes https://github.com/godotengine/godot-proposals/issues/522

Current look:
image
image
(EDIT: the screenshots don't match after I updated the one above, but you get the idea xd)

### CODE REVIEW

**Comment by KoBeWi** [comment]
Ok I implemented everything as originally requested.

**Comment by reduz** [comment]
I think this looks good, but don't we have this for the multiplayer testing already?

**Comment by KoBeWi** [comment]
Well, THIS is for multiplayer testing. Right now we just have an option to run multiple instances, as a dropdown.
image

**Review by Calinou — COMMENTED** [review]
Tested locally (rebased on top of `master` da81ca62a5f6d615516929896caa0b6b09ceccfc), it works as expected.

There are some things that should be improved before merging:

- The SpinBox that sets the number of instances should be disabled if **Enable Multiple Instances** is unchecked.
- **Override Main Run Args** should have a tooltip to denote that when unchecked, it will append to **Main Run Args** instead of overridding.
- The dialog can be made extremely tall if set to `20` instances. I'd suggest setting the upper limit to `12`, which fits on a 1080p display at 100% scaling.

Additional thought for a future PR:

- If you reduce the number of instances and increase it, options for instances that were removed are lost, even if you do this without closing and reopening the dialog. I suggest keeping removed instance options in memory at least until the dialog is closed (preferably until the editor is closed). This way, you can decrease/increase the number of instances without losing options for unused instances.
  - This makes me wonder if multi-instance options should be persisted to `project.godot` as well, just like **Main Run Args**.

**Comment by KoBeWi** [comment]
> This makes me wonder if multi-instance options should be persisted to project.godot as well, just like Main Run Args.

Main Run Args aren't only for debugging. They can be relevant for an exported project.

> If you reduce the number of instances and increase it, options for instances that were removed are lost, even if you do this without closing and reopening the dialog.

I could make the arguments persist across restarts by not truncating the saved array.

**Comment by Calinou** [comment]
> Main Run Args aren't only for debugging. They can be relevant for an exported project.

Reading the code, **Main Run Args** is only used when running a project from the editor. We currently don't have an equivalent that also works on an exported project, in case you want to override a CLI argument that has no equivalent project setting. (We should strive to add project settings for those instead, as it's more flexible overall.)

**Comment by KoBeWi** [comment]
Ah, I don't know then. Moving it to Project Settings is easy, but I'd do it once someone actually requests it. I think `main_run_args` setting only exists because previously it was not possible to specify it otherwise.

**Comment by fire** [comment]
https://gist.github.com/mrcdk/3c22d91a3b4fb784d5df60667a33cf1e

>  Godot 4.x example of how to launch multiple instances and tile them by using an editor debugger plugin to communicate with each instance 

from 

Mastadon:

https://mstdn.ca/@jdbaudi@mastodon.gamedev.place/110899576749166547

**Comment by KoBeWi** [comment]
This is some debugger hack. This PR makes implementing it much easier, because you can pass different arguments to instances.

**Comment by cyraid** [comment]
@KoBeWi Could you make it for Project Settings? I may change my mind with one project and do different CLI arguments, and then do different setup with another project. I wouldn't want to ruin the previous project just because it were Editor Settings. If that's what I'm understanding? Hoping it's Project Settings. ^_^

**Comment by KoBeWi** [comment]
It's not in editor settings, it's in project metadata, i.e. hidden per-project settings.

**Comment by cyraid** [comment]
@KoBeWi so this is from a year ago.. Does it really take this long for approvals? What is the main blocking point?

**Comment by KoBeWi** [comment]
It needs to be reviewed by a maintainer. Some PRs wait longer than others, because contributors are busy with other stuff and they get forgotten.

**Comment by KoBeWi** [comment]
Well anyone can review, but only reviews from @godotengine members count towards merging. idk if you can see it:
image
Reviews/approvals from non-members might be helpful when assessing the PR, but usually random users aren't qualified to make technical review (they can however test if the PR works correctly or whether there aren't any typos etc.).

> I try and give gentle reminder and a nice boost of optimism but Calinou shuts it down and says not to post unless you have any new information (in different issue #).

That's not how it works. Sure, bumping an issue might get it on someone's radar, but most often it just causes pointless notification, because if someone didn't have time to review something before, a random bump will not change that. The discussion on GitHub should be more focused; if you want to bring attention to the PR, you can come to the contributor chat. You can also leave a review as I mentioned. It has similar effect to a random comment, but has more value.

**Comment by BastiaanOlij** [comment]
Cool, @Calinou just pointed this to me as I was running into similar situations.

One thing I'd like to see added here is to not only be able to specify individual arguments for each instance, but also feature tags.

Bonus points if we can also start an instance on a remote device. I'd like to start a spectator server on desktop simultaneously with running my app on a Quest.

And would there be an ability to save this as presets, so I can switch between different configurations easily?

**Comment by cyraid** [comment]
@BastiaanOlij would the [Multirun Addon](https://github.com/xiezi5160/MultirunForGodot4) be of any use? Please submit an issue and tag me if you can explain there tagging/presets and I'll see if I can add it for ya.

**Comment by KoBeWi** [comment]
> One thing I'd like to see added here is to not only be able to specify individual arguments for each instance, but also feature tags.

See #63529 I'll integrate that once this PR is merged.

> Bonus points if we can also start an instance on a remote device.

No idea how this is done, maybe I could check the native run code 🤔 Not sure about argument support though.

> And would there be an ability to save this as presets, so I can switch between different configurations easily?

I'd leave that for later.

**Review by Calinou — APPROVED** [review]
Tested locally (rebased on top of https://github.com/godotengine/godot/commit/10e111477db68fe65776a1d68fb1ffccaf6520fc), it works as expected. The rebase has conflicts, but these are trivial to resolve.

Here are 12 simultaneous instances (the maximum you can set) in their full glory: 

image

**Comment by akien-mga** [comment]
Needs a rebase.

Overall I feel this dialog would still need a UX pass to make it look a bit nicer and with good use of vertical and horizontal space. And some of the other features we discussed above. But that can be done after this one is merged.

**Comment by akien-mga** [comment]
`editor_scale.h` moved to `editor/themes/`.

File: editor/run_instances_dialog.cpp
Comment (adamscott) [inline editor/run_instances_dialog.cpp:283]: There should be a `ScrollContainer` containing the `argument_container`, in order to make sure that it doesn't add children indefinitely.

Comment (KoBeWi) [reply editor/run_instances_dialog.cpp:283]: That's what instance limit of 12 is for. I didn't want to use ScrollContainer, because it makes the size undefined, but I guess I could just give it a minimum size 🤔
I will improve this in the follow-up, for now I think it's fine as is.

**Comment by adamscott** [comment]
> The dialog can be made extremely tall if set to 20 instances. I'd suggest setting the upper limit to 12, which fits on a 1080p display at 100% scaling.

I suggest that we don't instantiate a limit based on the UX. If so, it means that the UX is bad. I think that adding a scrollbar would ease the pain of having more than 12 instances.

File: editor/editor_run.cpp
Comment (AThousandShips) [inline editor/editor_run.cpp:241]:
```suggestion
			}
```

File: editor/run_instances_dialog.cpp
Comment (AThousandShips) [inline editor/run_instances_dialog.cpp:289]:
```suggestion
}

```

File: editor/run_instances_dialog.h
Comment (AThousandShips) [inline editor/run_instances_dialog.h:68]:
```suggestion
	Vector<String> _split_cmdline_args(const String &p_arg_string) const;
```

File: editor/run_instances_dialog.cpp
Comment (AThousandShips) [inline editor/run_instances_dialog.cpp:196]:
```suggestion
			if (exec_args.size() > 0) {
```
Is more clear IMO

**Comment by KoBeWi** [comment]
Added scroll container:
image
However the dialog can no longer shrink, so it's meh with low count:
image

**Pull request merged**
